
<h1 align="center"> Battle of the Neighbourhoods - Toronto </h1>

## Author: José D'Andrade

# Question 1

Scrape the page https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M and create a dataframe


In [1]:
import pandas as pd
import requests

In [2]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(url)
source

<Response [200]>

Response 200 means that we were able to make the connection to the page

<b>read_html reads all available tables. We are interested in table at index 0

In [3]:
wikitables = pd.read_html(source.text)
len(wikitables)

print(f'Total tables on page: {len(wikitables)}')

Total tables on page: 3


<b>The dataframe will consist of three columns: PostalCode, Borough, and Neighbourhood

In [4]:
table = wikitables[0]
table

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

<b>Only process the cells that have an assigned borough.
<b>Ignore cells with a borough that is Not assigned.

In [5]:
df = table[table["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

<b> If a cell has a borough but a Not assigned Neighbourhood, then it will be the same as the borough.
    
    1 Group the records based on Postal Code
    2 Get a count of Not Assigned Neighbourhood

In [6]:
df = df.groupby(['Postal Code']).head()
df.Neighbourhood.str.count("Not assigned").sum()

0

Nothing else to do here since there are no Not assigned Neighbourhood

<b>In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

In [7]:
df.shape

(103, 3)

We have 103 rows and 3 columns

# Question 2

As suggested in the assignment, given that geocoder can be unreliable, I am using the csv file that has the geographical coordinates of each postal code from http://cocl.us/Geospatial_data

<b> import the CSV file from the URL

In [8]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [9]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [10]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [11]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

**Dimensions are the same, so let's join on the postal codes to get the data.**

In [12]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data.shape

(103, 5)

**Use the Geocoder package or the csv file to create the dataframe:**

103 rows now displaying latitude and longitude

In [13]:
combined_data

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
2                                            Parkwoods  43.753259 -79.329656  
3                                     Victoria Village  43.725882 -79.315572  
4                            Regent Park, Harbourfront  43.654260 -79.360636  
5                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
6          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
160      The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
165                               Church and Wellesley  43.665860 -79.383160  
168  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
178  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

# Question 3

Explore and cluster the neighborhoods in Toronto. I chose to replicate the same analysis we did to the New York City data.

**download all the dependencies that we will need**

In [14]:
import numpy as np # library to handle data in a vectorized manner

import json # library to handle JSON files

from geopy.geocoders import Nominatim  # convert an address into latitude and longitude values

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


**explore the dataset and examine the dataframe**

In [15]:
combined_data.head()

Postal Code           Borough                                Neighbourhood  \
2         M3A        North York                                    Parkwoods   
3         M4A        North York                             Victoria Village   
4         M5A  Downtown Toronto                    Regent Park, Harbourfront   
5         M6A        North York             Lawrence Manor, Lawrence Heights   
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
2  43.753259 -79.329656  
3  43.725882 -79.315572  
4  43.654260 -79.360636  
5  43.718518 -79.464763  
6  43.662301 -79.389494

**how many boroughs and neighborhoods are there?**

In [16]:
print('The dataframe has {} boroughs and {} neighbourhoods.'.format(
        len(combined_data['Borough'].unique()),
        combined_data.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighbourhoods.


**use geopy library to get the latitude and longitude values of Toronto**

In [17]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


**create a map of Toronto with neighborhoods superimposed on top**

In [18]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

**initialize Foursquare API credentials**

**(LINE 19 REMOVED TO PROTECT PRIVATE INFORMATION)**

In [ ]:
CLIENT_ID = 'ABCDEFGHIJKLMNO12345678902MPBR53UIDGTXUV0F1KSO4A' 
CLIENT_SECRET = 'ABCDEFGHIJKLMNO1234567890GK3AHIQZRODW3Q5PJHR0TQB'
VERSION = '20180605' # Foursquare API version

**create a function to get all the venue categories in Toronto**

In [20]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name) # comment to avoid displaying all names
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

**collect the venues in Toronto for each Neighbourhood**

In [21]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [22]:
venues_in_toronto.shape

(1318, 5)

In [23]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

**group venues by Neighbourhood**

In [24]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1305  Mimico NW, The Queensway West, South of Bloor,...   
1306  Mimico NW, The Queensway West, South of Bloor,...   
1307  Mimico NW, The Queensway West, South of Bloor,...   
1308  Mimico NW, The Queensway West, South of Bloor,...   
1309  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1305               43.628841               -79.520999              Wingporium   
1306               43.628841               -79.520999        South St. Burger   
1307               43.628841               -79.520999               Dollarama   
1308               43.628841               -79.520999          Healthy Planet   
1309               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1305            Wings Joint  
1306           Burger Joint  
1307         Discount Store  
1308        Supplement Shop  
1309                 Bakery  

[409 rows x 5 columns]

Checking for the maximum venue categories

In [25]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
Airport Service     CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
Airport Service                  43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.379817   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                               Venue  
Venue Category                                        
Accessories Store                Ardene Shoes Outlet  
Airport              Toronto Downsview Airport (YZD)  
Airport Food Court                 Billy Bishop Café  
Airport Lounge                         Porter Lounge  
Airport Service     Porter Airlines Check-In Counter  
...                                              ...  
Warehouse Store                               Costco  
Wine Bar                           The National Club  
Wings Joint                               Wingporium  
Women's Store                          Maximum Woman  
Yoga Studio                                YogaSpace  

[237 rows x 4 columns]

**237 different types of Venue Categories**

### **Perform a One Hot encoding on venue categories**

In [26]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Lounge  \
0                     0        0                   0               0   
1                     0        0                   0               0   
2                     0        0                   0               0   
3                     0        0                   0               0   
4                     0        0                   0               0   
...                 ...      ...                 ...             ...   
1313                  0        0                   0               0   
1314                  0        0                   0               0   
1315                  0        0                   0               0   
1316                  0        0                   0               0   
1317                  0        0                   0               0   

      Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                   0                 0                    0             0   
1                   0                 0                    0             0   
2                   0                 0                    0             0   
3                   0                 0                    0             0   
4                   0                 0                    0             0   
...               ...               ...                  ...           ...   
1313                0                 0                    0             0   
1314                0                 0                    0             0   
1315                0                 0                    0             0   
1316                0                 0                    0             0   
1317                0                 0                    0             0   

      Aquarium  Art Gallery  ...  Train Station  Truck Stop  \
0            0            0  ...              0           0   
1            0            0  ...              0           0   
2            0            0  ...              0           0   
3            0            0  ...              0           0   
4            0            0  ...              0           0   
...        ...          ...  ...            ...         ...   
1313         0            0  ...              0           0   
1314         0            0  ...              0           0   
1315         0            0  ...              0           0   
1316         0            0  ...              0           0   
1317         0            0  ...              0           0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1313                              0                 0                      0   
1314                              0                 0                      0   
1315                              0                 0                      0   
1316                              0                 0                      0   
1317                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4                   0         0            0              0            0  
...               ...       ...     

**adding the neighbourhood to the encoded dataframe**

In [27]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Train Station  Truck Stop  \
0             0         0  ...              0           0   
1             0         0  ...              0           0   
2             0         0  ...              0           0   
3             0         0  ...              0           0   
4             0         0  ...              0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 238 columns]

**group the Neighbourhoods, and calculate the mean venue categories in each Neighbourhood**

In [28]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Lounge  Airport Service  \
0      0.0                 0.0             0.0              0.0   
1      0.0                 0.0             0.0              0.0   
2      0.0                 0.0             0.0              0.0   
3      0.0                 0.0             0.0              0.0   
4      0.0                 0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  ...  \
0               0.0                 0.00           0.0       0.0  ...   
1               0.0                 0.00           0.0       0.0  ...   
2               0.0                 0.00           0.0       0.0  ...   
3               0.0                 0.00           0.0       0.0  ...   
4               0.0                 0.04           0.0       0.0  ...   

   Train Station  Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  \
0            0.0         0.0                            0.0               0.0   
1            0.0         0.0                            0.0               0.0   
2            0.0         0.0                            0.0               0.0   
3            0.0         0.0                            0.0               0.0   
4            0.0         0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 238 columns]

Let's make a function to get the top most common venue categories

In [29]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [30]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village                       Café   
4                Bedford Park, Lawrence Manor East                Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge        Breakfast Spot          Skating Rink   
1                  Pool           Coffee Shop                   Gym   
2                  Bank                  Park    Chinese Restaurant   
3   Japanese Restaurant    Chinese Restaurant                  Bank   
4    Italian Restaurant        Sandwich Place         Grocery Store   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
1        Sandwich Place          Skating Rink                          Pub   
2           Bridal Shop        Sandwich Place                        Diner   
3           Yoga Studio           Escape Room            Electronics Store   
4            Restaurant     Indian Restaurant          Japanese Restaurant   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Drugstore            Donut Shop                Dog Run  
1                  Yoga Studio      Department Store           Dessert Shop  
2                   Restaurant          Intersection          Deli / Bodega  
3  Eastern European Restaurant             Drugstore             Donut Shop  
4                    Juice Bar          Liquor Store                Butcher

### **make the model to cluster Neighbourhoods**

In [31]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [32]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Checking the labelling of our model

In [33]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 3, 1, 1, 1, 1, 1, 2, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1,
       0, 1, 1, 1, 1, 1, 0])

Let's add the clustering Label column to the top 10 common venue categories

In [34]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [35]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
2         M3A        North York                                    Parkwoods   
3         M4A        North York                             Victoria Village   
4         M5A  Downtown Toronto                    Regent Park, Harbourfront   
5         M6A        North York             Lawrence Manor, Lawrence Heights   
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels  1st Most Common Venue  \
2  43.753259 -79.329656             0.0                   Park   
3  43.725882 -79.315572             1.0  Portuguese Restaurant   
4  43.654260 -79.360636             1.0            Coffee Shop   
5  43.718518 -79.464763             1.0         Clothing Store   
6  43.662301 -79.389494             1.0            Coffee Shop   

  2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
2     Food & Drink Shop            Yoga Studio      Department Store   
3           Coffee Shop           Hockey Arena     French Restaurant   
4                  Park                 Bakery        Breakfast Spot   
5     Accessories Store  Vietnamese Restaurant              Boutique   
6      Sushi Restaurant               Creperie              Beer Bar   

         5th Most Common Venue 6th Most Common Venue  \
2                  Escape Room     Electronics Store   
3  Eastern European Restaurant             Drugstore   
4                         Café           Yoga Studio   
5       Furniture / Home Store           Event Space   
6                Smoothie Shop        Sandwich Place   

         7th Most Common Venue       8th Most Common Venue  \
2  Eastern European Restaurant                   Drugstore   
3                   Donut Shop                     Dog Run   
4                   Restaurant                         Pub   
5                  Coffee Shop  Construction & Landscaping   
6                Burrito Place                        Café   

   9th Most Common Venue 10th Most Common Venue  
2             Donut Shop                Dog Run  
3          Deli / Bodega    Distribution Center  
4         Chocolate Shop           Dessert Shop  
5            Escape Room      Electronics Store  
6  Portuguese Restaurant     College Auditorium

Drop all the NaN values to prevent data skew

In [36]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [37]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [38]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

### Cluster 1

In [39]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2          North York             0.0                  Park   
32               York             0.0                  Park   
57          East York             0.0                  Park   
98               York             0.0                  Park   
100        North York             0.0                  Park   
135       Scarborough             0.0                  Park   
147  Downtown Toronto             0.0                  Park   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2       Food & Drink Shop           Yoga Studio      Department Store   
32          Women's Store         Deli / Bodega           Escape Room   
57      Convenience Store           Yoga Studio      Department Store   
98          Jewelry Store           Yoga Studio      Department Store   
100     Convenience Store           Yoga Studio      Department Store   
135            Playground          Intersection         Deli / Bodega   
147            Playground                 Trail          Dance Studio   

    5th Most Common Venue        6th Most Common Venue  \
2             Escape Room            Electronics Store   
32      Electronics Store  Eastern European Restaurant   
57            Escape Room            Electronics Store   
98            Escape Room            Electronics Store   
100           Escape Room            Electronics Store   
135     Electronics Store  Eastern European Restaurant   
147     Electronics Store  Eastern European Restaurant   

           7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
2    Eastern European Restaurant             Drugstore            Donut Shop   
32                     Drugstore            Donut Shop               Dog Run   
57   Eastern European Restaurant             Drugstore            Donut Shop   
98   Eastern European Restaurant             Drugstore            Donut Shop   
100  Eastern European Restaurant             Drugstore            Donut Shop   
135                    Drugstore            Donut Shop               Dog Run   
147                    Drugstore            Donut Shop               Dog Run   

    10th Most Common Venue  
2                  Dog Run  
32     Distribution Center  
57                 Dog Run  
98                 Dog Run  
100                Dog Run  
135    Distribution Center  
147    Distribution Center

### Cluster 2

In [40]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels  1st Most Common Venue  \
3          North York             1.0  Portuguese Restaurant   
4    Downtown Toronto             1.0            Coffee Shop   
5          North York             1.0         Clothing Store   
6    Downtown Toronto             1.0            Coffee Shop   
9         Scarborough             1.0   Fast Food Restaurant   
..                ...             ...                    ...   
157  Downtown Toronto             1.0             Restaurant   
160         Etobicoke             1.0                  River   
165  Downtown Toronto             1.0            Coffee Shop   
168      East Toronto             1.0     Light Rail Station   
178         Etobicoke             1.0          Grocery Store   

    2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
3             Coffee Shop           Hockey Arena     French Restaurant   
4                    Park                 Bakery        Breakfast Spot   
5       Accessories Store  Vietnamese Restaurant              Boutique   
6        Sushi Restaurant               Creperie              Beer Bar   
9             Yoga Studio            Event Space           Escape Room   
..                    ...                    ...                   ...   
157                  Café            Coffee Shop    Seafood Restaurant   
160            Smoke Shop                   Pool          Dance Studio   
165            Hobby Shop            Escape Room                  Café   
168           Pizza Place             Skate Park               Brewery   
178        Hardware Store      Convenience Store        Discount Store   

           5th Most Common Venue        6th Most Common Venue  \
3    Eastern European Restaurant                    Drugstore   
4                           Café                  Yoga Studio   
5         Furniture / Home Store                  Event Space   
6                  Smoothie Shop               Sandwich Place   
9              Electronics Store  Eastern European Restaurant   
..                           ...                          ...   
157                    Gastropub             Asian Restaurant   
160  Eastern European Restaurant                    Drugstore   
165                          Pub             Ramen Restaurant   
168                Burrito Place               Farmers Market   
178                 Burger Joint               Sandwich Place   

    7th Most Common Venue       8th Most Common Venue  9th Most Common Venue  \
3              Donut Shop                     Dog Run          Deli / Bodega   
4              Restaurant                         Pub         Chocolate Shop   
5             Coffee Shop  Construction & Landscaping            Escape Room   
6           Burrito Place                        Café  Portuguese Restaurant   
9               Drugstore                  Donut Shop                Dog Run   
..                    ...                         ...                    ...   
157                   Pub                   Speakeasy                  Hotel   
160            Donut Shop                     Dog Run    Distribution Center   
165        Ice Cream Shop           Indian Restaurant               Beer Bar   
168  Fast Food Restaurant                  Restaurant          Auto Workshop   
178  Fast Food Restaurant             Supplement Shop                 Bakery   

    10th Most Common Venue  
3      Distribution Center  
4             Dessert Shop  
5        Electronics Store  
6       College Auditorium  
9      Distribution Center  
..                     ...  
157             Steakhouse  
160         Discount Store  
165   Ethiopian Restaurant  
168                    Spa  
178          Tanning Salon  

[88 rows x 12 columns]

### Cluster 3

In [41]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
89   North York             2.0  Construction & Landscaping   
169   Etobicoke             2.0              Baseball Field   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
89         Baseball Field           Yoga Studio          Dessert Shop   
169           Yoga Studio    Falafel Restaurant  Ethiopian Restaurant   

    5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
89   Ethiopian Restaurant           Escape Room            Electronics Store   
169           Escape Room     Electronics Store  Eastern European Restaurant   

           8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
89   Eastern European Restaurant             Drugstore             Donut Shop  
169                    Drugstore            Donut Shop                Dog Run

### Cluster 4

In [42]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
129  Central Toronto             3.0           Summer Camp   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
129           Yoga Studio           Event Space           Escape Room   

    5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
129     Electronics Store  Eastern European Restaurant             Drugstore   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
129            Donut Shop               Dog Run    Distribution Center

### Cluster 5

In [43]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
17  Etobicoke             4.0   Filipino Restaurant           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
17           Event Space           Escape Room     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
17  Eastern European Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
17               Dog Run    Distribution Center